klippa/src/hvar.rs

```diff
@@ -0,0 +1,640 @@
+    }
+
+    fn to_serialize_plan(&self) -> DeltaSetIndexMapSerializePlan {
+        DeltaSetIndexMapSerializePlan::new(
+            self.outer_bit_count,
+            self.inner_bit_count,
+            &self.output_map,
+            self.map_count as u32,
+        )
+    }
+}
+
+#[derive(Default)]
+pub(crate) struct HvarVvarSubsetPlan {
+    inner_maps: Vec<IncBiMap>,
+    index_map_subset_plans: Vec<IndexMapSubsetPlan>,
+}
+
+impl HvarVvarSubsetPlan {
+    pub(crate) fn new(
+        plan: &Plan,
+        var_store: &ItemVariationStore,
+        index_maps: &[Option<DeltaSetIndexMap>],
+    ) -> Result<Self, ReadError> {
+        let mut this = HvarVvarSubsetPlan::default();
+        let vardata_count = var_store.item_variation_data_count() as usize;
+
+        let mut inner_sets = Vec::new();
+        inner_sets.resize(vardata_count, Default::default());
+
+        let mut outer_map = IncBiMap::default();
+        // don't bypass empty adv map
+        let index_map_subset_plan =
+            IndexMapSubsetPlan::init(&index_maps[0], plan, false, &mut outer_map, &mut inner_sets)?;
+        this.index_map_subset_plans.push(index_map_subset_plan);
+
+        let mut adv_set = IntSet::empty();
+        if index_maps[0].is_none() {
+            adv_set.union(&inner_sets[0]);
+        }
+
+        for index_map in index_maps.iter().skip(1) {
+            this.index_map_subset_plans.push(IndexMapSubsetPlan::init(
+                index_map,
+                plan,
+                true,
+                &mut outer_map,
+                &mut inner_sets,
+            )?);
+        }
+
+        outer_map.sort();
+
+        let retain_adv_map = index_maps[0].is_none()
+            && plan
+                .subset_flags
+                .contains(SubsetFlags::SUBSET_FLAGS_RETAIN_GIDS);
+
+        if retain_adv_map {
+            let mut inner_map = IncBiMap::with_capacity(plan.new_to_old_gid_list.len());
+            for (_, old_gid) in plan.new_to_old_gid_list.iter() {
+                inner_map.add(old_gid.to_u32());
+            }
+            this.inner_maps.push(inner_map);
+        } else {
+            let mut inner_map = IncBiMap::default();
+            for old_gid in adv_set.iter() {
+                inner_map.add(old_gid as u32);
+            }
+            inner_sets[0].subtract(&adv_set);
+
+            for i in inner_sets[0].iter() {
+                inner_map.add(i as u32);
+            }
+
+            this.inner_maps.push(inner_map);
+        }
+
+        for set in inner_sets.iter().skip(1) {
+            let mut map = IncBiMap::default();
+            for i in set.iter() {
+                map.add(i as u32);
+            }
+            this.inner_maps.push(map);
+        }
+
+        for (i, index_map) in index_maps.iter().enumerate() {
+            this.index_map_subset_plans[i].remap(plan, index_map, &outer_map, &this.inner_maps);
+        }
+        Ok(this)
+    }
+
+    pub(crate) fn inner_maps(&self) -> &Vec<IncBiMap> {
```

Comment:
unless there's a good reason (i.e. you need `&mut Vec<T>`) you should generally prefer to pass `&[T]` to `&Vec<T>`.